src/Native/Runtime/arm64/GcProbe.S

```diff
@@ -0,0 +1,212 @@
+    //  maskConstant  : 1 if savedRegsMask is a constant value
+    .macro INIT_PROBE_FRAME threadReg, trashReg, savedRegsMask, gcFlags, frameSize, maskConstant
+         str         \threadReg, [sp, #OFFSETOF__PInvokeTransitionFrame__m_pThread]            // Thread *
+
+        .if \maskConstant == 1
+            movz    \trashReg, (\savedRegsMask & 0xFFFF)
+            movk    \trashReg, (\savedRegsMask >> 16) & 0xFFFF, lsl 16
+            movk    \trashReg, (\gcFlags) & 0xFFFF, lsl 32
+            movk    \trashReg, (\gcFlags >> 16) & 0xFFFF, lsl 48
+            str     \trashReg, [sp, #OFFSETOF__PInvokeTransitionFrame__m_Flags]
+        .else
+            str     \savedRegsMask, [sp, #OFFSETOF__PInvokeTransitionFrame__m_Flags]
+        .endif
+
+        add         \trashReg, sp, #\frameSize
+        str         \trashReg, [sp, #m_CallersSP]
+    .endm
+
+    .macro PROLOG_PROBE_FRAME threadReg, trashReg, savedRegsMask, gcFlags, getThread = 0, maskConstant = 0
+
+        // Define the method prolog, allocating enough stack space for the PInvokeTransitionFrame and saving
+        // incoming register values into it.
+        ALLOC_PROBE_FRAME 0, 1
+
+        .if \getThread == 1
+#if FEATURE_EMULATED_TLS
+            bl RhpGetThread
+
+            .ifnc \threadReg, x0
+                mov \threadReg, x0
+            .endif
+#else
+            INLINE_GETTHREAD \threadReg
+#endif
+        .endif
+
+        // Perform the rest of the PInvokeTransitionFrame initialization.
+        INIT_PROBE_FRAME \threadReg, \trashReg, \savedRegsMask, \gcFlags, PROBE_FRAME_SIZE, \maskConstant
+        mov         \trashReg, sp
+        str         \trashReg, [\threadReg, #OFFSETOF__Thread__m_pHackPInvokeTunnel]
+    .endm
+
+    // Simple macro to use when PROLOG_PROBE_FRAME was used to set up and initialize the prolog and
+    // PInvokeTransitionFrame. This will define the epilog including a return via the restored LR.
+    .macro EPILOG_PROBE_FRAME
+
+        FREE_PROBE_FRAME 0, 1
+        EPILOG_RETURN
+    .endm
+
+//
+// Set the Thread state and wait for a GC to complete.
+//
+// Register state on entry:
+//  x4: thread pointer
+//  
+// Register state on exit:
+//  x4: thread pointer
+//  All other registers trashed
+//
+
+    .global RhpWaitForGCNoAbort
+
+    .macro WaitForGCCompletion
+
+        ldr         w2, [x4, #OFFSETOF__Thread__m_ThreadStateFlags]
+        tst         w2, #TSF_SuppressGcStress__OR__TSF_DoNotTriggerGC
+        bne         0f
+
+        ldr         x9, [x4, #OFFSETOF__Thread__m_pHackPInvokeTunnel]
+        bl          RhpWaitForGCNoAbort
+0:
+    .endm
+
+    LEAF_ENTRY RhpGcPoll
+        PREPARE_EXTERNAL_VAR_INDIRECT_W RhpTrapThreads, 0
+        tbnz        x0, #TrapThreadsFlags_TrapThreads_Bit, RhpGcPollRare
+        ret
+    LEAF_END RhpGcPoll
+
+    NESTED_ENTRY RhpGcPollRare, _TEXT, NoHandler
+        PROLOG_PROBE_FRAME x2, x3, PROBE_SAVE_FLAGS_EVERYTHING, 0, 1, 1
```

Comment:
It should be enough to use `PUSH_COOP_PINVOKE_FRAME` for this. We should not need to save all registers that `PROLOG_PROBE_FRAME` saves. (And if this needed to save all register, we would need to preserve all registers in `RhpGcPoll` too.)